ambry-store/src/main/java/com.github.ambry.store/PersistentIndex.java

```diff
@@ -712,29 +753,38 @@ IndexValue markAsPermanent(StoreKey id, FileSpan fileSpan, long operationTimeMs)
    * @return the {@link IndexValue} of the ttl update record
    * @throws StoreException if there is any problem writing the index record
    */
-  private IndexValue markAsPermanent(StoreKey id, FileSpan fileSpan, MessageInfo info, long operationTimeMs)
-      throws StoreException {
+  private IndexValue markAsPermanent(StoreKey id, FileSpan fileSpan, MessageInfo info, long operationTimeMs,
+      short lifeVersion) throws StoreException {
     validateFileSpan(fileSpan, true);
+    boolean hasLifeVersion = lifeVersion != -1;
+    //TODO IndexValue findLatest
     IndexValue value = findKey(id);
     if (value == null && info == null) {
       throw new StoreException("Id " + id + " not present in index " + dataDir, StoreErrorCodes.ID_Not_Found);
     } else if (value != null && value.isFlagSet(IndexValue.Flags.Delete_Index)) {
       throw new StoreException("Id " + id + " deleted in index " + dataDir, StoreErrorCodes.ID_Deleted);
     } else if (value != null && value.isFlagSet(IndexValue.Flags.Ttl_Update_Index)) {
       throw new StoreException("TTL of " + id + " already updated in index" + dataDir, StoreErrorCodes.Already_Updated);
+    } else if (hasLifeVersion && value.getLifeVersion() > lifeVersion) {
```

Comment:
same as above

Reply:
@justinlin-linkedin    

lifeVersion >= value.getLifeVersion() is allowed.  Reference: https://docs.google.com/document/d/1rm6NBOUeZRMbuF26116lerytiW3tde38I23cZVr3SG0/edit#heading=h.hdj8rn37deoq